### PR TITLE
8266550: C2: mirror TypeOopPtr/TypeInstPtr/TypeAryPtr with TypeKlassPtr/TypeInstKlassPtr/TypeAryKlassPtr

### PR DESCRIPTION
**Pull request by rwestrel**
This is some refactoring in another attempt to fix JDK-6312651
(Compiler should only use verified interface types for
optimization). Rather than propose the patch from:

https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2019-May/033803.html

as a single big patch. I've been working on splitting it. The plan is
to have this and another refactoring patch that include no change to
the way interfaces are handled as preparation. Then only, in a third
patch, interface support along the lines of the patch I proposed 2
years ago would be introduces.

This patch changes the class hierarchy of types that C2 uses and
introduces TypeInstKlassPtr/TypeAryKlassPtr that mirror
TypeInstPtr/TypeAryPtr. The motivation for this is that a single:

ciKlass* _klass;

is no longer sufficient to describe a type (a set of interfaces must
also be carried around). That's not possible with TypeKlassPtr.

The meet methods for TypeInstPtr/TypeInstKlassPtr and
TypeAryPtr/TypeAryKlassPtr are largely similar. I moved the most
complicated logic in helper methods:

meet_instptr() and meet_aryptr()

(Thanks to Vladimir I for testing the refactoring patches)

/cc hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8266550](https://bugs.openjdk.java.net/browse/JDK-8266550): C2: mirror TypeOopPtr/TypeInstPtr/TypeAryPtr with TypeKlassPtr/TypeInstKlassPtr/TypeAryKlassPtr

### Reviewers
 * [Vladimir Ivanov](https://openjdk.java.net/census#vlivanov) (@iwanowww - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3880/head:pull/3880` \
`$ git checkout pull/3880`

Update a local copy of the PR: \
`$ git checkout pull/3880` \
`$ git pull https://git.openjdk.java.net/jdk pull/3880/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3880`

View PR using the GUI difftool: \
`$ git pr show -t 3880`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3880.diff">https://git.openjdk.java.net/jdk/pull/3880.diff</a>

</details>

### CODE REVIEW

**Comment by rwestrel** [comment]
rebased to latest jdk code. Anyone for this?

**Comment by iwanowww** [comment]
I started reviewing the patch.

Can you, please, elabroate on design goals of the patch? 
I don't fully understand the intended relation between `TypeInstKlassPtr`, `TypeAryKlassPtr`, and `TypeKlassPtr`.

My observations so far:

* `Type::base()` can never be `Type::KlassPtr` anymore: it's either `InstKlassPtr` or `AryKlassPtr`; all `t->base() == Type::KlassPtr` should be replaced by `t->base() == Type::InstKlassPtr || t1->base() == Type::AryKlassPtr`;
 
* `java.lang.Object` is represented with `InstKlassPtr` (`TypeInstKlassPtr::OBJECT`) and as a consequence type flattening can turn `TypeAryKlassPtr` into `TypeInstKlassPtr`.

**Comment by rwestrel** [comment]
> I started reviewing the patch.

Thanks!
> 
> Can you, please, elabroate on design goals of the patch?
> I don't fully understand the intended relation between `TypeInstKlassPtr`, `TypeAryKlassPtr`, and `TypeKlassPtr`.

On the oop side, TypeOopPtr is the super type for TypeInstPtr and TypeAryPtr. Type*KlassPtr are intended to work the same.
To keep track of interfaces, a TypeInstPtr will need an extra field that contains the set of interfaces that it implements.
A TypeAryPtr implements some interfaces (Cloneable etc.). If its elements are objects, its elements may implement interfaces too. A TypeAryPtr has a TypeAry field that references a TypeOopPtr for object arrays. Interfaces implemented by the array elements are attached to that TypeOopPtr.

Now with only a TypeKlassPtr where do we record interfaces implemented by array elements?

The motivation of this change is to mirror the structure described above with TypeOopPtr on the TypeKlassPtr and to have somewhere to store interfaces for object arrays.

> My observations so far:
> 
> * `Type::base()` can never be `Type::KlassPtr` anymore: it's either `InstKlassPtr` or `AryKlassPtr`; all `t->base() == Type::KlassPtr` should be replaced by `t->base() == Type::InstKlassPtr || t1->base() == Type::AryKlassPtr`;

I don't think that's true: roughly TypeAryKlassPtr meets TypeInstKlassPtr would give a TypeKlassPtr (same as TypeAryPtr meets TypeInstPtr gives a TypeOopPtr).

> * `java.lang.Object` is represented with `InstKlassPtr` (`TypeInstKlassPtr::OBJECT`) and as a consequence type flattening can turn `TypeAryKlassPtr` into `TypeInstKlassPtr`.

That should again be the same as what's going on with TypeOopPtr/TypeInstPtr/TypeAryPtr.

**Comment by iwanowww** [comment]
> TypeAryKlassPtr meets TypeInstKlassPtr would give a TypeKlassPtr (same as TypeAryPtr meets TypeInstPtr gives a TypeOopPtr).

Can you point me to the relevant code, please? What I'm seeing so far is that, in constrast to `TypeOopPtr`, `TypeKlassPtr` is never instantiated at runtime and  `TypeKlassPtr` always points to either `TypeInstKlassPtr` or `TypeAryKlassPtr`.

**Comment by rwestrel** [comment]
> > TypeAryKlassPtr meets TypeInstKlassPtr would give a TypeKlassPtr (same as TypeAryPtr meets TypeInstPtr gives a TypeOopPtr).
> 
> Can you point me to the relevant code, please? What I'm seeing so far is that, in constrast to `TypeOopPtr`, `TypeKlassPtr` is never instantiated at runtime and `TypeKlassPtr*` always points to either `TypeInstKlassPtr` or `TypeAryKlassPtr`.

You're right about that (my reply was incorrect). I don't understand why it matters this said. TypeKlassPtr is for pointers to classes that are not known to be a instance class or` an array class. That might prove useful at some point.

**Comment by iwanowww** [comment]
>  I don't understand why it matters this said. TypeKlassPtr is for pointers to classes that are not known to be a instance class or` an array class. That might prove useful at some point.

There are leftover usages of `Type::KlassPtr` in the code. In particular:
```
src/hotspot/share/opto//idealGraphPrinter.cpp:      } else if (t->base() == Type::KlassPtr) {
src/hotspot/share/opto//memnode.cpp:  } else if (tp->base() == Type::KlassPtr) {
```

Also, `Type::category()` is not adjusted for `Type::InstKlassPtr`/`Type::AryKlassPtr`.

**Comment by iwanowww** [comment]
Another case is in `Compile::flatten_alias_type()`:
```
  // Flatten all to bottom for now
  switch( _AliasLevel ) {
  ...
  case 1:                       // Flatten to: oop, static, field or array
    switch (tj->base()) {
    case Type::RawPtr:   tj = TypeRawPtr::BOTTOM;   break;
    case Type::AryPtr:   // do not distinguish arrays at all
    case Type::InstPtr:  tj = TypeInstPtr::BOTTOM;  break;
    case Type::KlassPtr: tj = TypeInstKlassPtr::OBJECT; break;
    case Type::AnyPtr:   tj = TypePtr::BOTTOM;      break;  // caller checks it
```

`Type::KlassPtr` should be accompanied by `Type::InstKlassPtr` and `Type::AryKlassPtr` cases, shouldn't it?

**Comment by rwestrel** [comment]
> `Type::KlassPtr` should be accompanied by `Type::InstKlassPtr` and `Type::AryKlassPtr` cases, shouldn't it?

Good catch. I fix the ones you found. They seem mostly harmless, this said.

File: src/hotspot/share/opto/callnode.cpp [hunk 1]
```diff
@@ -377,9 +377,6 @@ static void format_helper( PhaseRegAlloc *regalloc, outputStream* st, Node *n, c
     case Type::InstPtr:
       st->print(" %s%d]=#Ptr" INTPTR_FORMAT,msg,i,p2i(t->isa_oopptr()->const_oop()));
       break;
-    case Type::KlassPtr:
```
Comment (iwanowww) [inline src/hotspot/share/opto/callnode.cpp:380]: Why do you remove the case? I'd expect to see 2 new cases added: 
```
   case Type::KlassPtr:
+  case Type::AryKlassPtr:
+  case Type::InstKlassPtr:
```

File: src/hotspot/share/opto/compile.cpp [hunk 2]
```diff
@@ -1453,7 +1453,8 @@ const TypePtr *Compile::flatten_alias_type( const TypePtr *tj ) const {
     case Type::RawPtr:   tj = TypeRawPtr::BOTTOM;   break;
     case Type::AryPtr:   // do not distinguish arrays at all
     case Type::InstPtr:  tj = TypeInstPtr::BOTTOM;  break;
-    case Type::KlassPtr: tj = TypeKlassPtr::OBJECT; break;
+    case Type::AryKlassPtr:
```
Comment (iwanowww) [inline src/hotspot/share/opto/compile.cpp:1456]: Similar situation here. Why don't you leave `Type::KlassPtr` intact and just extend it to `Type::AryKlassPtr` and `Type::InstKlassPtr` cases?

**Review by iwanowww — APPROVED** [review]
Overall, it looks good.

Testing results (hs-tier1 - hs-tier6) are fine. 

Just a couple of minor comments follow.

**Comment by rwestrel** [comment]
@iwanowww Thanks for the review (and help along the way). I updated the change with your suggestions.

File: src/hotspot/share/opto/node.cpp [hunk 3]
```diff
@@ -1787,10 +1787,7 @@ void Node::dump(const char* suffix, bool mark, outputStream *st) const {
   if (t != NULL && (t->isa_instptr() || t->isa_klassptr())) {
     const TypeInstPtr  *toop = t->isa_instptr();
     const TypeKlassPtr *tkls = t->isa_klassptr();
-    ciKlass*           klass = toop ? toop->klass() : (tkls ? tkls->klass() : NULL );
-    if (klass && klass->is_loaded() && klass->is_interface()) {
-      st->print("  Interface:");
-    } else if (toop) {
+    if (toop) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/node.cpp:1790]: Why did you remove the interface case?

File: src/hotspot/share/opto/type.hpp [hunk 4]
```diff
@@ -1799,14 +1879,32 @@ inline const TypeMetadataPtr *Type::isa_metadataptr() const {
...
+  assert( _base == InstKlassPtr, "Not a klass pointer" );
+  return (TypeInstKlassPtr*)this;
+}
+
+inline const TypeAryKlassPtr *Type::isa_aryklassptr() const {
+  return (_base == AryKlassPtr) ? (TypeAryKlassPtr*)this : NULL;
+}
+
+inline const TypeAryKlassPtr *Type::is_aryklassptr() const {
+  assert( _base == AryKlassPtr, "Not a klass pointer" );
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.hpp:1904]: The extra whitespaces should be removed.

File: src/hotspot/share/opto/type.hpp [hunk 5]
```diff
@@ -1799,14 +1879,32 @@ inline const TypeMetadataPtr *Type::isa_metadataptr() const {
...
+  assert( _base >= KlassPtr && _base <= AryKlassPtr, "Not a klass pointer" );
   return (TypeKlassPtr*)this;
 }
 
+inline const TypeInstKlassPtr *Type::isa_instklassptr() const {
+  return (_base == InstKlassPtr) ? (TypeInstKlassPtr*)this : NULL;
+}
+
+inline const TypeInstKlassPtr *Type::is_instklassptr() const {
+  assert( _base == InstKlassPtr, "Not a klass pointer" );
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.hpp:1895]: The extra whitespaces should be removed.

File: src/hotspot/share/opto/type.hpp [hunk 6]
```diff
@@ -1799,14 +1879,32 @@ inline const TypeMetadataPtr *Type::isa_metadataptr() const {
 }
 
 inline const TypeKlassPtr *Type::isa_klassptr() const {
-  return (_base == KlassPtr) ? (TypeKlassPtr*)this : NULL;
+  return (_base >= KlassPtr && _base <= AryKlassPtr ) ? (TypeKlassPtr*)this : NULL;
 }
 
 inline const TypeKlassPtr *Type::is_klassptr() const {
-  assert( _base == KlassPtr, "Not a klass pointer" );
+  assert( _base >= KlassPtr && _base <= AryKlassPtr, "Not a klass pointer" );
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.hpp:1886]: The extra whitespaces should be removed.

File: src/hotspot/share/opto/type.cpp [hunk 7]
```diff
@@ -5116,42 +5527,56 @@ ciKlass* TypeAryPtr::klass() const {
...
+  if (elem->isa_klassptr() && !klass_is_exact) {
+    elem = elem->is_klassptr()->cast_to_exactness(klass_is_exact);
+  }
+  return make(klass_is_exact ? Constant : NotNull, elem, k, _offset);
 }
 
 
 //-----------------------------as_instance_type--------------------------------
 // Corresponding type for an instance of the given class.
 // It will be NotNull, and exact if and only if the klass type is exact.
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.cpp:5568]: Is this comment still valid?

File: src/hotspot/share/opto/library_call.cpp [hunk 8]
```diff
@@ -5968,7 +5968,7 @@ bool LibraryCallKit::inline_cipherBlockChaining_AESCrypt(vmIntrinsics::ID id) {
 
   ciInstanceKlass* instklass_AESCrypt = klass_AESCrypt->as_instance_klass();
   const TypeKlassPtr* aklass = TypeKlassPtr::make(instklass_AESCrypt);
-  const TypeOopPtr* xtype = aklass->as_instance_type();
+  const TypeOopPtr* xtype = aklass->as_instance_type()->cast_to_ptr_type(TypePtr::NotNull);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/library_call.cpp:5971]: I assume you need this because you've changed `as_instance_type` to no longer cast to `TypePtr::NotNull`. Why is that and what about other usages of `as_instance_type`?

File: src/hotspot/share/opto/type.hpp [hunk 9]
```diff
@@ -916,6 +924,19 @@ class TypePtr : public Type {
   void dump_inline_depth(outputStream *st) const;
 #endif
 
+  enum MeetResult {
+    QUICK,
+    UNLOADED,
+    SUBTYPE,
+    NOT_SUBTYPE,
+    LCA
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/type.hpp:932]: Please add comments explaining the meaning of these.

**Review by TobiHartmann — COMMENTED** [review]
Hard to review but looks good to me. Just added some minor comments/questions.

File: src/hotspot/share/opto/node.cpp (hunk 3 above)
Comment (rwestrel) [reply src/hotspot/share/opto/node.cpp]: I brought it back as this doesn't really matter for this PR. The next change in the series removes the klass() accessor and that creates all sort of issues. In some cases where the use of klass() seems to have little value (like here, after all interfaces can't be trusted), I removed the code using the klass() rather than try to fix it.

File: src/hotspot/share/opto/type.hpp (hunk 4 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.hpp]: done.

File: src/hotspot/share/opto/type.hpp (hunk 5 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.hpp]: done.

File: src/hotspot/share/opto/type.hpp (hunk 6 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.hpp]: done.

File: src/hotspot/share/opto/type.cpp (hunk 7 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.cpp]: Nice catch. Fixed.

File: src/hotspot/share/opto/library_call.cpp (hunk 8 above)
Comment (rwestrel) [reply src/hotspot/share/opto/library_call.cpp:5971]: I looked at other use of as_instance_type and found that there was not need for the cast to not null. As to why, it makes little sense to me that as_instance_type() would return NotNull. It feels error prone.

File: src/hotspot/share/opto/type.hpp (hunk 9 above)
Comment (rwestrel) [reply src/hotspot/share/opto/type.hpp:937]: I added a comment. Let me know if it's ok.

**Comment by rwestrel** [comment]
> Hard to review but looks good to me. Just added some minor comments/questions.

Thanks for reviewing this.

**Review by TobiHartmann — APPROVED** [review]
Thanks for making these changes. Looks good to me.

**Comment by rwestrel** [comment]
I had to tweak test/hotspot/jtreg/compiler/lib/ir_framework/IRNode.java after merging with master. @iwanowww  @TobiHartmann  are you still ok with the change?

**Review by iwanowww — APPROVED** [review]
Still looks good to me.

**Review by TobiHartmann — APPROVED** [review]
Still looks good to me.

**Comment by rwestrel** [comment]
@TobiHartmann @iwanowww thanks for the re-review!
